Add docs and template ProviderDataIngester (#790)
WordPress/openverse-catalog: openverse_catalog/docs/provider_data_ingester_faq.md

# ProviderDataIngester FAQ

The most straightforward implementation of a `ProviderDataIngester` repeats the following process:

* Builds a set of query params for the next request, based upon the previous params (for example, by updating offsets or page numbers)
* Makes a single GET request to the configured static `endpoint` using the built params
* Extracts a "batch" of records from the response, as a list of record JSON representations
* Iterates over the records in the batch, extracting desired data, and commits them to local storage

Some provider APIs may not fit neatly into this workflow. This document addresses some common use cases.

## How do I process a provider "record" that contains data about multiple records?

**Example**: You're pulling data from a Museum database, and each "record" in a batch contains multiple photos of a single physical object.

> **Solution**: The `get_record_data` method takes a `data` object representing a single record from the provider API. Typically, it extracts required data and returns it as a single dict. However, it can also return a **list of dictionaries** for cases like the one described, where multiple Openverse records can be extracted.

Reviewer: It would be nice to add links to the examples of each of the solutions.

Author: I wanted to do this but was worried about code drift and making this document difficult to maintain 🤔

Reviewer 2: Ooo, that's a really good point!

Reviewer: The problem with code drift and maintainability is really a good point :( I wonder if we could automate it somehow, add a link to some regex search string that would lead to the example of, say, overridden method in one of the solutions...  

Not necessary for this PR, but would be really nice to have. I usually learn by example and links to the examples of the solutions would be really helpful for the way _I_ learn :)

```
def get_record_data(self, data: dict) -> dict | list[dict] | None:
    records = []

    for record in data.get("associatedImages", []):
        # Perform data processing for each record and add it to a list
        records.append({
            "foreign_landing_url": record.get("foreign_landing_url")
            ...
        })

    return records
```

## What if I can't get all the necessary information for a record from a single API request?

**Example**: A provider's `search` endpoint returns a list of image records containing *most* of the information we need for each record, but not image dimensions. This information *is* available via the API by hitting a `details` endpoint for a given image, though.

**Solution**: In this case, you can reuse the `get_response_json` method by passing in the endpoint you need:

> ```

Author: Excellent point, that's a huge improvement! Collapsing the rest of these into a single commit :)

def get_record_data(self, data: dict) -> dict | list[dict] | None:
    ...

    # Get data from the details endpoint
    response_json = self.get_response_json(
        query_params={"my_params": "foo"},
        endpoint=f"https://foobar.museum.org/api/v1/images/{data.get("uuid")}"
    )

    ...
```

> When doing this, keep in mind that adding too many requests may slow down ingestion. Be aware of rate limits from your provider API as well.

Reviewer 2: This is a really important point, so I think extra emphasis is worth it:
```suggestion
**NOTE**: When doing this, keep in mind that adding too many requests may slow down ingestion. Be aware of rate limits from your provider API as well.
```

## What if my API endpoint isn't static and needs to change from one request to another?

> Example: Rather than passing a `page` number in query parameters, a provider expects the `page` as part of the endpoint path itself.

Reviewer 2: ```suggestion
**Example**: Rather than passing a `page` number in query parameters, a provider expects the `page` as part of the endpoint path itself.
```

> If your `endpoint` needs to change, you can implement it as a `property`:

Reviewer 2: ```suggestion
**Solution**: If your `endpoint` needs to change, you can implement it as a `property`:
```

```python
@property
def endpoint(self) -> str:
    # Compute the endpoint using some instance variable
    return f"https://foobar.museum.org/images/page/{self.page_number}"
```

In this example, `self.page_number` is an instance variable that gets updated after each request. To set up the instance variable you can override `__init__`, **being careful to remember to call `super` and pass through kwargs**, and then update it in `get_next_query_params`:

```python
def __init__(self, *args, **kwargs):

>     # IMPORTANT!

Reviewer 2: 🚲 🏠 for sure, but maybe this would convey the point a bit more explicitly?
```suggestion
    # REQUIRED!
```

    super().__init__(*args, **kwargs)

    # Set up our instance variable
    self.page_number = None

def get_next_query_params(self, prev_query_params: dict | None, **kwargs) -> dict:
    # Remember that `get_next_query_params` is called before every request, even
    # the first one.

    if self.page_number is None:
        # Set initial value
        self.page_number = 0
    else:
        # Increment value on subsequent requests
        self.page_number += 1

    # Return your actual query params
    return {}
```

Now each time `get_batch` is called, the `endpoint` is correctly updated.

## How do I run ingestion for a set of discrete categories?

> **Example**: My provider has some set of categories that I'd like to iterate over and ingest data for. Eg, an audio provider's search endpoint that requires you specify whether you're searching for "podcasts", "music", etc. I'd like to iterate over all the available categories and run ingestion for each.

Reviewer 2: Nit:
```suggestion
**Example**: My provider has some set of categories that I'd like to iterate over and ingest data for. E.g., an audio provider's search endpoint requires you specify whether you're searching for "podcasts", "music", etc. I'd like to iterate over all the available categories and run ingestion for each.
```

Reviewer 2: Ah! That's fair 🙂
